CRM/Membershipextrasimporterapi/EntityImporter/LineItem.php

```diff
@@ -0,0 +1,303 @@
+    }
+  }
+
+  private function getQuantity() {
+    if (!empty($this->rowData['line_item_quantity'])) {
+      return $this->rowData['line_item_quantity'];
+    }
+
+    return 1;
+  }
+
+  private function getUnitPrice() {
+    return $this->rowData['line_item_unit_price'];
+  }
+
+  private function getLineTotal($unitPrice, $quantity) {
+    return MoneyUtilities::roundToCurrencyPrecision($unitPrice * $quantity);
+  }
+
+  private function getFinancialTypeId() {
+    if (!isset($this->cachedValues['financial_types'])) {
+      $sqlQuery = "SELECT id, name FROM civicrm_financial_type";
+      $result = CRM_Core_DAO::executeQuery($sqlQuery);
+      while ($result->fetch()) {
+        $this->cachedValues['financial_types'][$result->name] = $result->id;
+      }
+    }
+
+    if (!empty($this->cachedValues['financial_types'][$this->rowData['line_item_financial_type']])) {
+      return $this->cachedValues['financial_types'][$this->rowData['line_item_financial_type']];
+    }
+
+    throw new CRM_Membershipextrasimporterapi_Exception_InvalidLineItemException('Invalid line item "Financial Type"', 200);
+  }
+
+  private function createFinancialItemRecord($lineItemId, $mappedLineItemParams) {
+    $sqlParams = [
+      1 => [$this->contribution['contact_id'], 'Integer'],
+      2 => [$mappedLineItemParams['line_item_label'], 'String'],
+      3 => [$mappedLineItemParams['line_total'], 'Money'],
+      4 => [$this->contribution['currency'], 'String'],
+      5 => [$this->getToFinancialAccountId($mappedLineItemParams['financial_type_id']), 'Integer'],
+      6 => [$this->getFinancialItemStatusId(), 'Integer'],
+      7 => ['civicrm_line_item', 'String'],
+      8 => [$lineItemId, 'Integer'],
+      9 => [$this->contribution['receive_date'], 'String'],
+    ];
+    $sqlQuery = "INSERT INTO `civicrm_financial_item` (`contact_id` , `description` , `amount` , `currency` ,
+                 `financial_account_id` , `status_id` , `entity_table` , `entity_id`, `transaction_date`) 
+                 VALUES (%1, %2, %3, %4, %5, %6, %7, %8, %9)";
+    CRM_Core_DAO::executeQuery($sqlQuery, $sqlParams);
+
+    $dao = CRM_Core_DAO::executeQuery('SELECT LAST_INSERT_ID() as id');
+    $dao->fetch();
+    return $dao->id;
+  }
+
+  private function getToFinancialAccountId($financialTypeId) {
+    // this is reserved value and will always equal such value on any CiviCRM site
+    $incomeAccountRelationshipId = 1;
+
+    $sqlQuery = "SELECT financial_account_id FROM civicrm_entity_financial_account 
+                   WHERE entity_table = 'civicrm_financial_type' AND entity_id = {$financialTypeId} AND account_relationship = {$incomeAccountRelationshipId}";
+    $result = CRM_Core_DAO::executeQuery($sqlQuery);
+    $result->fetch();
+    return $result->financial_account_id;
+  }
+
+  private function getFinancialItemStatusId() {
+    $contributionStatusId = $this->contribution['contribution_status_id'];
+
+    // hardcoded Ids are used for efficiency reasons
```

Comment:
It is okay to hardcode it, but can you give the meaning of the 1, 9, 2, 5, 8 in the comment perhaps?  Do not need to declare variables for them.

Reply:
/g/hardcoded/Hardcoded